### DISCUSSION TITLE
Does "containerized socket activation" improve security?

### DISCUSSION THREAD

**Discussion by eriksjolund** [containers/podman#13390] [Q&A] [answered]
The live video streaming and chat server software [owncast](https://github.com/owncast/owncast/) makes use of two listening TCP sockets (one for the web server and one for RTMP).   Using _socket activation_ for multiple different sockets requires the use of the environment variable `LISTEN_FDNAMES` to distinguish the different of sockets from each other. I wanted to try it out, so I created a  [feature request](https://github.com/owncast/owncast/issues/1750) and  wrote a [PR](https://github.com/owncast/owncast/pull/1751). The functionality seems to work. (I've only verified it by checking that the web server provides a web page when accessing it).

Unfortunately the feature request was closed with the motivation that there is no need for the requested feature.
Now I wonder if I maybe should have put more emphasis on the security advantage of using containerized socket activation.
(At least to me there seems to be a security advantage).

This was one of the rationale arguments I provided in the feature request:

_If you have a containerized server that takes requests from the network via a listening TCP port, adding socket activation support opens up the possibility to run the container with the --network=none command-line option. This will restrict the container from accessing the network except via the TCP socket it has been passed by systemd via Podman. Outgoing connections would not be allowed. Running the container with --network=none improves security. I'm not sure if this is relevant in the case of owncast though._

(I hope that was a correct description of how it works)

Has anyone given any thoughts to how "containerized socket activation" relates to security? Any pros and cons? 

A side-note regarding SELinux:
[container-selinux](https://github.com/containers/container-selinux) needs to be at least version __v2.179.0__
to avoid having to pass the option  `--security-opt label=disable` to `podman run`. 

By the way, what is a suitable name for the functionality?
"_containerized socket activation_" or maybe better  "_socket activation to container_" or something else?

**Comment by rhatdan** [comment, marked as answer]
I like the way you talk about this from a security point of view.  The --network=none with socket activation seems like a great idea.

I could see this being useful with a display only webserver for example.  Where you could allow incoming connections not not allow connections back out.  But this would only work with TCP connections.

**Reply by eriksjolund** [reply]
@rhatdan Thanks for the feedback!

> But this would only work with TCP connections.

I'm not sure how to parse that sentence. Could you explain a bit more? Do you mean TCP in comparison to Unix sockets, or maybe TCP in comparison to UDP?

**Reply by rhatdan** [reply]
Yes TCP, and Unix Domain Sockets would work, but not udp, I believe.

**Reply by eriksjolund** [reply]
I made a quick test with `ListenDatagram=127.0.0.1:3000` and a socket activated container. It worked with __--network=none__.

I should go through the test more carefully to see that I have not done any mistake.

**Reply by eriksjolund** [reply]
New plan: I'll publish a minimal demo about using UDP but it might take a few weeks.

**Comment by rhatdan** [comment]
If you could just give a few examples of each use case, we could write a cool blog on it.
Unix Domain Socket, TCP Socket, UDP Socket.

**Comment by rhatdan** [comment]
If you have examples of each, I will write up the Blog.

**Reply by eriksjolund** [reply]
Sure, that would be cool! Right now I have an echo server that upcase the string it receives 

**TCP**
```
$ echo hello | socat - tcp:127.0.0.1:3000
HELLO
```

**UDP**
```
$  echo hello | socat -t 5 - udp:127.0.0.1:3000
HELLO
```
I'll send you something during the weekend.

**Comment by rhatdan** [comment]
Yes a demo showing a leaked socket working and then exec into the container and showing that you can not connect out using curl or point ...

**Reply by eriksjolund** [reply]
@rhatdan See
https://github.com/eriksjolund/asio/blob/systemd_socket_activation/socket_activation.md

(I hope to clean the code up a bit). I wrote you an email too (from erik.sjolund@gmail.com).

**Reply by rhatdan** [reply]
Thanks, this is looking really good.  I updated containers-selinux to allow socket activation for tcp_sockets and udp_sockets from systemd and users including unconfined_t.

**Reply by eriksjolund** [reply]
Thanks, after upgrading to __container-selinux-2.181.0-1.fc35.noarch__ from https://bodhi.fedoraproject.org
`systemd-socket-activate` now works without `--security-opt label=disable`!

Some changes:
I've

* deleted https://github.com/eriksjolund/asio and instead created the new repository https://github.com/eriksjolund/socket-activate-echo where I put the source code.
* added a [GitHub Actions workflow](https://github.com/eriksjolund/socket-activate-echo/blob/main/.github/workflows/publish_container_image.yml) that builds the multi-arch container image  [__ghcr.io/eriksjolund/socket-activate-echo__](https://github.com/eriksjolund/socket-activate-echo/pkgs/container/socket-activate-echo) (archs: linux/amd64, linux/arm64, linux/riscv64, linux/ppc64le, linux/s390x, linux/arm/v7)
* modified the documentation and the service file to make use of that container image.
* added a troubleshooting section where I discussed the __socat__ timeout configuration and the possibility of using `--pull=never` and pulling the image beforehand.

**Comment by rhatdan** [comment]
Excellent, now time to throw together a blog on running containers more securely by turning off network access, but using systemd socket activation to allow limited network connectivity.

**Reply by eriksjolund** [reply]
An extra idea for the blog post:

* mentioning that _socket activation_ is supported in two places in podman:
   -  when starting the Podman API service   
   -  when starting a container
   
  I've seen some confusion about this in some internet forum posts, where it was believed that the socket activated API support would also imply the other type of support.


Something unrelated:

I'm curious if it would make sense to add SELinux support for `AF_VSOCK` socket activation?

I made a test with a slightly patched echo server and

```
$ cat echo@demo.socket.d/override.conf
[Socket]
ListenStream=vsock:1:3000
```

The echo worked
```
$ echo hello | socat - VSOCK-CONNECT:1:3000
hello
```
audit2allow then showed

```
allow container_t container_runtime_t:vsock_socket { accept getattr getopt };
```

When I tested with `systemd-socket-activate`, audit2allow showed

```
allow container_t unconfined_t:vsock_socket { accept getattr getopt };
```

I could create a proper GitHub issue at https://github.com/containers/container-selinux/issues
with the ausearch raw output, if you see any use case for AF_VSOCK.
(~I don't know much about VMs. Would it be possible to pass an AF_VSOCK file descriptor via podman/conmon/crun/libkrun to a container?~)

**Reply by eriksjolund** [reply]
> I'm curious if it would make sense to add SELinux support for AF_VSOCK socket activation?

Answering my own question after thinking a bit more about it:

Yes, I think there could be some use cases, for instance, running the echo container in a systemd user service in a Fedora CoreOS VM and then connecting to it from the host with __socat__. (I assume AF_VSOCK provides better performance than AF_INET)

I created two feature requests in the __container-selinux__ repository:

https://github.com/containers/container-selinux/issues/175
https://github.com/containers/container-selinux/issues/176

The AF_VSOCK functionality resides in the git branch _vsock_:
https://github.com/eriksjolund/socket-activate-echo/tree/vsock

The tag _vsock_ is used for the container image:
 ghcr.io/eriksjolund/socket-activate-echo:vsock

**Reply by eriksjolund** [reply]
I merged
https://github.com/eriksjolund/socket-activate-echo/tree/vsock
into the main branch. (AF_VSOCK SOCK_STREAM works with container-selinux 2.183.0)

**Reply by eriksjolund** [reply]
@rhatdan I haven't heard from you regarding the blog post so I started writing some text.
A draft: https://github.com/eriksjolund/podman-socket-activation
If the text somehow could be useful for a https://www.redhat.com/sysadmin/
blog post (a whole blog post or just a part of a blog post), please let me know.

**Reply by rhatdan** [reply]
Excellent, I have been a bit distracted.

**Reply by rhatdan** [reply]
I like the blog. One thing I would show is podman exec into the container without network connectivity, after you show it working, and then attempt to talk on the network.  Show that it is blocked.  Explain that this is the equivalent  to a hacked web server being used to launch attacks against other hosts.

**Reply by eriksjolund** [reply]
Thanks, good points. I'll update it. Probably during the weekend.

**Reply by rhatdan** [reply]
Were do you want to publish this?  Do you want to publish it locally or are you interested in publishing somewhere else like redhat.com/sysadmin?

**Reply by eriksjolund** [reply]
I would be happy to publish it in redhat.com/sysadmin. There would be more readers and it would also be something that I could add to my resume.

**Reply by rhatdan** [reply]
Ok I can help you do that.

**Reply by eriksjolund** [reply]
Thanks!

**Reply by eriksjolund** [reply]
Instead of redhat.com/sysadmin, do you think it would be possible to add the document under
https://github.com/containers/podman/tree/main/docs/tutorials
and https://docs.podman.io/en/latest/Tutorials.html
?
(I thought maybe the document will then be used and reused more)
I created  a [branch](https://github.com/eriksjolund/podman-socket-activation/tree/rework_text) where I made the document to be more of a tutorial and less of a  blog. (I understand a proper review process would need to follow before the document could land upstream). The document is not quite ready yet but for the most part.

**Reply by rhatdan** [reply]
I would prefer it in both.  sysadmin gets eyeballs. While adding it to tutorials would be useful for others to find it in the future. 

One would be written from a security point of view for sysadmin, the other would be to explain how to use socket activation in containers.

**Reply by eriksjolund** [reply]
SGTM, I'll split it into two. I'l try to do it over the weekend.

**Reply by eriksjolund** [reply]
I split it into 3 documents
 
1. https://github.com/eriksjolund/podman-socket-activation/blob/rework_text/blog.md
     Draft blog text for redhat/sysadmin

2. https://github.com/eriksjolund/podman-socket-activation/blob/rework_text/README.md
   Possible starting point for a podman.io tutorial.

3. https://github.com/eriksjolund/podman-socket-activation/blob/rework_text/RestrictAddressFamilies.md
    Probably in the unsupported-use-of-Podman-territory, so I guess I could put the text somewhere locally

**Comment by eriksjolund** [comment]
I created another example 
https://github.com/eriksjolund/socket-activate-httpd
(A web server that can't establish outgoing connections)

Yes, I agree, time for a blog. Tell me if you would like help.

**Comment by eriksjolund** [comment]
#### Edit 1

~Another reflection regarding using podman and socket activation:~

~The data communication in the socket does not go through slirp4netns when running rootless Podman.~
~Hopefully this means better network performance (throughput).~

~I haven't done any benchmarking to verify this theory though.~

I opened up a new discussion thread about the network performance

#### Edit 2: fixed link 2 September 2024

[_Does "containerized socket activation" improve network performance?_](https://github.com/containers/podman/discussions/14068)

**Reply by zeronumbers** [reply]
I believe you wrote wrong link since it points to this very page where we are now, the one you wanted to link is probably: https://github.com/containers/podman/discussions/14068

I didn't read whole discussion, and am still learning what this socket stuff is, but I guess another important question is whether "containerized socket activation" could be exploited in any way, or there are literally no downsides (for security) guaranteed?

**Reply by eriksjolund** [reply]
Thanks, for catching this. I've now fixed the link.

Here are some other links. I wrote two blog posts:

* [_How to limit container privilege with socket activation_](https://www.redhat.com/sysadmin/socket-activation-podman)
* [_How to restrict network access in Podman with systemd_](https://www.redhat.com/sysadmin/podman-systemd-limit-access)


> I didn't read whole discussion, and am still learning what this socket stuff is, but I guess another important question is whether "containerized socket activation" could be exploited in any way, or there are literally no downsides (for security) guaranteed?



SELinux didn't allow containerized socket activation at first, but after relaxing the rules somewhat, it now works.
 
* https://github.com/containers/container-selinux/issues/171
* https://github.com/containers/container-selinux/issues/172
* https://github.com/containers/container-selinux/issues/174
* https://github.com/containers/container-selinux/issues/175
* https://github.com/containers/container-selinux/issues/176
* https://github.com/containers/container-selinux/issues/179

Maybe this could be seen as a downside, although I don't see it that way.
